### PR TITLE
feat: change the writable glossary by a right click

### PR DESCRIPTION
**Pull request by Gen-Rainbow**
Add a new feature. Now users can change the writable glossary by right clicking the editor window. (When the glossary folder has more than one files)
屏幕截图 2024-08-22 091042

## Pull request type

- Feature enhancement -> [enhancement]

## Which ticket is resolved?

- 932: To add a function of selecting the glossary upon adding a word
  * https://sourceforge.net/p/omegat/feature-requests/932/
 
- 1003: Add the ability to select a glossary to add a glossary term to
  * https://sourceforge.net/p/omegat/feature-requests/1003/

## What does this PR change?

- src/org/omegat/gui/editor/EditorPopups.java
- src/org/omegat/Bundle.properties
- src/org/omegat/Bundle_zh_CN.properties

### CODE REVIEW

File: src/org/omegat/gui/editor/EditorPopups.java [hunk 1]
```diff
@@ -273,18 +275,71 @@ public void actionPerformed(ActionEvent e) {
             }
 
             menu.addSeparator();
+        }
+    }
+
+    /**
+     * Allow users to add new words and select writable glossary
+     *
+     * @author HanQin Chen
```
Comment (miurahr) [inline src/org/omegat/gui/editor/EditorPopups.java:284]: You can claim your self in line 69

**Review by miurahr — CHANGES_REQUESTED** [review]
There are several coding style issues. Please check my comments.

**Comment by miurahr** [comment]
> Warning: [ant:checkstyle] [WARN] /home/runner/work/omegat/omegat/src/org/omegat/gui/editor/EditorPopups.java:310: Line is longer than 120 characters (found 127). [LineLength]
> Warning: [ant:checkstyle] [WARN] /home/runner/work/omegat/omegat/src/org/omegat/gui/editor/EditorPopups.java:319: Line is longer than 120 characters (found 122). [LineLength]
> Warning: [ant:checkstyle] [WARN] /home/runner/work/omegat/omegat/src/org/omegat/gui/editor/EditorPopups.java:400: Line is longer than 120 characters (found 122). [LineLength]
> Warning: [ant:checkstyle] [WARN] /home/runner/work/omega/omegat/src/org/omegat/gui/editor/EditorPopups.java:437: Line is longer than 120 characters (found 121). [LineLength]
> 
There are also several checkstyle warnings in the change. Spotless will automatically format the code.

File: src/org/omegat/gui/editor/EditorPopups.java [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
-               2015 Aaron Madlon-Kay
-               2023 Thomas Cordonnier
-               Home page: https://www.omegat.org/
-               Support center: https://omegat.org/support
+ 2010 Wildrich Fourie
+ 2011 Alex Buloichik, Didier Briel
+ 2015 Aaron Madlon-Kay
+ 2023 Thomas Cordonnier
+ Home page: https://www.omegat.org/
+ Support center: https://omegat.org/support
```
[the remark this answers is not in the document]
Comment (Gen-Rainbow) [reply src/org/omegat/gui/editor/EditorPopups.java]: Sorry about that, I've reformatted the code according to the xml file.

**Comment by brandelune** [comment]
Please describe how your feature works. Add screenshots and precise descriptions please.

File: src/org/omegat/gui/editor/EditorPopups.java [hunk 3]
```diff
@@ -106,14 +114,18 @@ public void addItems(JPopupMenu menu, final JTextComponent comp, int mousepos,
...
-            // Use the project's target tokenizer to determine the word that was right-clicked.
-            // EditorUtils.getWordEnd() and getWordStart() use Java's built-in BreakIterator
-            // under the hood, which leads to inconsistent results when compared to other spell-
+            // Use the project's target tokenizer to determine the word that was
+            // right-clicked.
+            // EditorUtils.getWordEnd() and getWordStart() use Java's built-in
+            // BreakIterator
+            // under the hood, which leads to inconsistent results when compared
+            // to other spell-
             // checking functionality in OmegaT.
```
Comment (miurahr) [inline src/org/omegat/gui/editor/EditorPopups.java:123]:
```suggestion
            // EditorUtils.getWordEnd() and getWordStart() use Java's built-in
            // BreakIterator under the hood, which leads to inconsistent 
            // results when compared to other spell-checking functionality
            // in OmegaT.
```

**Review by miurahr — CHANGES_REQUESTED** [review]
Thank you for update the coding style. Here is a proposal to improve comment.

**Review by miurahr — CHANGES_REQUESTED** [review]
When switching writable glossary file, and adding a term to there in a team project, remote glossary is not updated. And when reloading, or close and open the project, the term, added by the feature, is gone. The local glossary file is over-write with a remote one.

When adding default glossary file, there is no problem.

Please fix the problem if you claim to add a feature to update non-default glossary, it also should work also with team projects. 

As a definition of OmegaT team project, non-standard glossary files are managed by team administrator and distributed by admin.
If you add a feature to allow users, not admin, to manipulate non-default glossaries,
you also need to design some mechanisms for a team project feature.

**Comment by Gen-Rainbow** [comment]
> When switching writable glossary file, and adding a term to there in a team project, remote glossary is not updated. And when reloading, or close and open the project, the term, added by the feature, is gone. The local glossary file is over-write with a remote one.
> 
> When adding default glossary file, there is no problem.
> 
> Please fix the problem if you claim to add a feature to update non-default glossary, it also should work also with team projects.
> 
> As a definition of OmegaT team project, non-standard glossary files are managed by team administrator and distributed by admin. If you add a feature to allow users, not admin, to manipulate non-default glossaries, you also need to design some mechanisms for a team project feature.

Well, that's something new to me. May need some time to figure it out.

**Comment by miurahr** [comment]
> I use git as my versioning tool and I have run a few tests. Everything worked well. Both me and my team members tried updating the non-standard glossary files and saving them (ctrl+S) and reloading the project (F5), they didn't just gone. The entries stay in the files which they belong to.

 "reloading the project(F5)" does not load remote content but reload only local content in OmegaT 6.0.0.
OmegaT 6.1.0, a master branch, it reloads contents from the remote. 

I close the project. and open it again when I observe it.

**Comment by miurahr** [comment]
Your change does not update a value of project properties of writable glossary path.
Your UI tells "change writable glossary", it is not consistent.

It is because OmegaT project has a writeable glossary property.

As in defined in `ProjectUICommand` class, a writable glossary file is defined as like 

```java
    public static void openWriteableGlossaryFile(boolean parent) {
        if (!Core.getProject().isProjectLoaded()) {
            return;
        }
        String path = Core.getProject().getProjectProperties().getWriteableGlossary();
        if (StringUtil.isEmpty(path)) {
            return;
        }
        File toOpen = new File(path);
        if (parent) {
            toOpen = toOpen.getParentFile();
        }
        openFile(toOpen);
    }
```

The method is called when you access OmegaT menu  "Project" > "Access Project Files" > "Access Writeable Glossary".
You may be interested in `ProjectProperties#setWriteableGlossary` method.

**Comment by Gen-Rainbow** [comment]
About the problem you pointed out, I think I might get it.

It seems that OmegaT will detect the writable glossary file when updating the project. If any changes were in the writable glossary file, it will be committed to the remote. In other words, changes in any other non-writable glossary file will just be ignored.

**Comment by Gen-Rainbow** [comment]
So if we want any changes in a non-default glossary file committed to the remote, that specific glossary file need to be set as the 'writable' glossary file.

**Comment by Gen-Rainbow** [comment]
> Your change does not update a value of project properties of writable glossary path.
> Your UI tells "change writable glossary", it is not consistent.

Please check the project properties (ctrl+E) again when you change the writable glossary. It should have been altered.

**Comment by Gen-Rainbow** [comment]
You may consider this feature as a shortcut to open the project props dialog then browser and change it manually (without reloading the project).

File: src/org/omegat/gui/editor/EditorPopups.java [hunk 4]
```diff
@@ -273,18 +286,88 @@ public void actionPerformed(ActionEvent e) {
...
+                String writableGlossPath = folder + s;
+                String test = props.getWriteableGlossary();
+                if (writableGlossPath.equals(test)) {
+                    item.setSelected(true);
+                }
+
+                item.addActionListener(new ActionListener() {
+                    @Override
+                    public void actionPerformed(ActionEvent e) {
+                        props.setWriteableGlossary(writableGlossPath);
```
Comment (miurahr) [inline src/org/omegat/gui/editor/EditorPopups.java:347]: Ok, I find it.

**Comment by brandelune** [comment]
> So if we want any changes in a non-default glossary file committed to the remote, that specific glossary file need to be set as the 'writable' glossary file.

I’m not sure that’s something we want to do.

The writeable glossary is the shared glossary.
Specific glossaries do not have to be shared glossaries. If fact, specific glossaries have good reasons to not be shared glossaries.

So before you go any further, I’d like you to precisely define what you want to have so that we can discuss what is good for OmegaT.

Thank you for your understanding.
Jean-Christophe
(OmegaT project manager)

**Comment by Gen-Rainbow** [comment]
> I’m not sure that’s something we want to do.
> 
> The writeable glossary is the shared glossary. Specific glossaries do not have to be shared glossaries. If fact, specific glossaries have good reasons to not be shared glossaries.
> 
> So before you go any further, I’d like you to precisely define what you want to have so that we can discuss what is good for OmegaT.
> 
> Thank you for your understanding. Jean-Christophe (OmegaT project manager)

I see. To be honest, I initially designed this feature because I encountered some inconvenience while using OmegaT alone. When I visited the OmegaT official website for help, I found that some users had the same problem as me. So I didn't consider the issue of team sharing, but directly created this feature.

I think maybe we can add a switch or something in the preference which give the right of turning on/off this feature to the users. I just tried to offer a different option.

For individual translators like myself, we can make use of this function. While translators work as a team can also just ignore it and don't need to change anything. It's all up to the users, all up to OmegaT project team.

**Review by miurahr — APPROVED** [review]

**Comment by brandelune** [comment]
Can we have a full description of the feature?

**Comment by miurahr** [comment]
We can specify a writable glossary file name from a project preference dialog, shown by `Project > Project  Preference`.
The change here provides a pop-up dialog to allow the user to select a writable glossary from currently placed glossary files.

**Comment by brandelune** [comment]
> We can specify a writable glossary file name from a project preference dialog, shown by `Project > Project Preference`. The change here provides a pop-up dialog to allow the user to select a writable glossary from currently placed glossary files.

@Gen-Rainbow Could you write a documentation ticket on Sourceforge with me as the owner for this documentation item?

**Pull request merged**